grug/malachite/src/actors/host/state.rs

```diff
@@ -0,0 +1,218 @@
+        Self {
+            db_storage: Box::new(storage),
+            consensus: None,
+            height: <ctx!(Height)>::new(1),
+            round: Round::Nil,
+            proposer: None,
+            role: Role::None,
+            config,
+            mempool_config,
+            started_round: Instant::now(),
+            app,
+            pending_commit_block_hash: None,
+        }
+    }
+
+    pub fn set_consensus(&mut self, consensus: ConsensusRef<Context>) {
+        self.consensus = Some(consensus);
+    }
+
+    pub fn consensus(&self) -> ActorResult<ConsensusRef<Context>> {
+        self.consensus
+            .clone()
+            .ok_or(anyhow::anyhow!("Consensus not set").into())
+    }
+
+    pub fn stream_id(&self) -> StreamId {
+        let mut bytes = Vec::with_capacity(size_of::<u64>() + size_of::<u32>());
+        bytes.extend_from_slice(&self.height.to_be_bytes());
+        // TODO: can this panic?
+        bytes.extend_from_slice(&self.round.as_u32().unwrap().to_be_bytes());
+        StreamId::new(bytes.into())
+    }
+
+    pub fn started_round(
+        &mut self,
+        height: ctx!(Height),
+        round: Round,
+        proposer: ctx!(Address),
+        role: Role,
+    ) {
+        self.started_round = Instant::now();
+        self.height = height;
+        self.round = round;
+        self.proposer = Some(proposer);
+        self.role = role;
+    }
+
+    pub fn calculate_block_sleep(&self) -> Duration {
+        let diff = self.started_round.elapsed();
+        if diff < self.config.block_time {
+            self.config.block_time - diff
+        } else {
+            Duration::ZERO
+        }
+    }
+
+    pub fn prepare_proposal(&self, txs: Vec<RawTx>) -> Vec<RawTx> {
+        self.app
+            .prepare_proposal(txs, self.mempool_config.max_txs_bytes.as_u64() as usize)
+    }
+
+    pub fn finalize_block<T>(&mut self, block: &Block<T>) -> AppResult<BlockHash> {
+        let app_hash = self.app.finalize_block(block.as_block_info(), &block.txs)?;
+
+        let block_hash = block.calculate_block_hash(app_hash);
+
+        self.pending_commit_block_hash = Some(block_hash);
+
+        Ok(block_hash)
+    }
+
+    pub fn commit(&mut self, block: &Block, consensus_batch: Batch) -> AppResult<()> {
+        let mut rerun = true;
+
+        // If the pending_commit_block_hash is equal to the block hash, this mean that the app batch in memory
+        // is the same as the block we are trying to commit, so we don't need to rerun the finalize_block.
+        if let Some(pending) = self.pending_commit_block_hash {
+            if pending == block.block_hash() {
+                rerun = false;
+            }
+        }
+
+        if rerun {
+            let block_hash = self.finalize_block(block)?;
+            if block_hash != block.block_hash() {
+                // TODO: We should panic here? This mean the the we have some non-deterministic behavior in the app.
+                panic!("Block hash mismatch");
```

Comment:
Not sure about the `pending_commit_block_hash` use here as a trigger for rerunning `finalize_block`. It is possible to get different proposals for same or different rounds. These are all finalized in earlier steps and information is stored in `UNDECIDED_BLOCK` and `UNDECIDED_PROPOSALS`, with the latter including the validity which is basically the check of the block_hash. 
So the caller can look in `UNDECIDED_PROPOSALS` and call `commit` with a valid block.
Added related comments in `decided()`.